### PR TITLE
RavenDB-13819 Move JS Index compilation cache from static level to database level

### PR DESCRIPTION
**Pull request by ppekrol**

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Static/IndexCompilationCache.cs [hunk 1]
```diff
@@ -8,31 +8,71 @@
...
-    /// we are going to have a single cache for all indexes. This also plays nice with testing, which 
-    /// will build up and tear down a server frequently, so we can still reduce the cost of compiling 
-    /// the indexes.
-    /// </summary>
-    public static class IndexCompilationCache
+    public class IndexCompilationCache
     {
-        private static readonly ConcurrentDictionary<CacheKey, Lazy<StaticIndexBase>> IndexCache = new ConcurrentDictionary<CacheKey, Lazy<StaticIndexBase>>();
+        /// <summary>
+        /// This is a static because creating indexes is expensive, we want to cache them
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/IndexCompilationCache.cs]: This isn't static, though

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Static/IndexCompilationCache.cs]: In general, the commend here is wrong

File: src/Raven.Server/Documents/Indexes/MapReduce/Static/MapReduceIndex.cs [hunk 2]
```diff
@@ -53,14 +53,17 @@ protected MapReduceIndex(MapReduceIndexDefinition definition, StaticIndexBase co
...
         public override bool IsMultiMap => _compiled.Maps.Count > 1 || _compiled.Maps.Any(x => x.Value.Count > 1);
 
-        public override void ResetIsSideBySideAfterReplacement()
+        protected override void DisposeIndex()
         {
-            _isSideBySide = null;
+            base.DisposeIndex();
+
+            if (Type.IsJavaScript())
+                DocumentDatabase?.IndexStore?.Cache.Remove(_compiled.CacheKey);
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/MapReduce/Static/MapReduceIndex.cs]: If you now remove it from the cache on shutdown, what use is the cache in the first place?

File: src/Raven.Server/Documents/Indexes/Static/MapIndex.cs [hunk 3]
```diff
@@ -53,6 +53,14 @@ protected MapIndex(MapIndexDefinition definition, StaticIndexBase compiled)
 
         public override bool IsMultiMap => _compiled.Maps.Count > 1 || _compiled.Maps.Any(x => x.Value.Count > 1);
 
+        protected override void DisposeIndex()
+        {
+            base.DisposeIndex();
+
+            if (Type.IsJavaScript())
+                DocumentDatabase?.IndexStore?.Cache.Remove(_compiled.CacheKey);
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/MapIndex.cs]: Should we just remove the cache completely?

File: src/Raven.Server/Documents/Indexes/Static/IndexCompilationCache.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/Indexes/Static/IndexCompilationCache.cs]: ups... this one should be `static`

File: src/Raven.Server/Documents/Indexes/Static/IndexCompilationCache.cs [hunk 4]
```diff
@@ -17,7 +17,7 @@ public class IndexCompilationCache
         /// will build up and tear down a server frequently, so we can still reduce the cost of compiling 
         /// the indexes.
         /// </summary>
-        private readonly ConcurrentDictionary<IndexCacheKey, Lazy<StaticIndexBase>> _staticIndexCache = new ConcurrentDictionary<IndexCacheKey, Lazy<StaticIndexBase>>();
+        private static readonly ConcurrentDictionary<IndexCacheKey, Lazy<StaticIndexBase>> _staticIndexCache = new ConcurrentDictionary<IndexCacheKey, Lazy<StaticIndexBase>>();
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/IndexCompilationCache.cs]: But now we are back at making this static globally, no? So what is the purpose?

Comment (ppekrol) [reply src/Raven.Server/Documents/Indexes/Static/IndexCompilationCache.cs]: This one is for C# indexes. Second one is for JS ones. I'm checking all usages of cache now, maybe we can remove JS ones from cache entirely

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Static/IndexCompilationCache.cs]: The main reason we have a cache for C# is to control the # of assemblies.
For JS, we probably can drop it

Comment (ppekrol) [reply src/Raven.Server/Documents/Indexes/Static/IndexCompilationCache.cs]: Done. Thx!

**Review by ayende — APPROVED** [review]

**Pull request merged**